File: apps/desktop/src/services/commands/notifications.ts
import { invoke } from "@/services/tauri";

export type ReminderKind = "daily_goal" | "due_review" | "streak" | "test";
export type NotificationOsStatus =
  | "sent"
  | "denied"
  | "unavailable"
  | "failed";

export interface NotificationSettings {
  userId: number;
  notificationEnabled: boolean;
  inAppRemindersEnabled: boolean;
  dueReviewNotificationsEnabled: boolean;
  streakNotificationsEnabled: boolean;
  reminderTime: string;
  reminderDaysOfWeek: string;
  updatedAt: string;
}

export interface UpdateNotificationSettingsInput {
  notificationEnabled: boolean;
  inAppRemindersEnabled: boolean;
  dueReviewNotificationsEnabled: boolean;
  streakNotificationsEnabled: boolean;
  reminderTime: string;
  reminderDaysOfWeek?: string;
}

export interface EvaluateRemindersInput {
  localDate?: string;
  localTime?: string;
  force?: boolean;
}

export interface InAppReminder {
  id: number;
  kind: ReminderKind;
  title: string;
  body: string;
  actionLabel?: string | null;
  route?: string | null;
  createdAt: string;
}

export interface ReminderEvaluation {
  reminders: InAppReminder[];
  newReminders: InAppReminder[];
  dueReviewCount: number;
  dailyGoalCards: number;
  todayCardsReviewed: number;
  currentStreak: number;
  settings: NotificationSettings;
  evaluatedAt: string;
}

export interface NotificationDispatchResult {
  reminder: InAppReminder;
  osStatus: NotificationOsStatus | "unavailable";
  message: string;
}

export const defaultNotificationSettings: NotificationSettings = {
  userId: 0,
  notificationEnabled: true,
  inAppRemindersEnabled: true,
  dueReviewNotificationsEnabled: true,
  streakNotificationsEnabled: true,
  reminderTime: "08:00",
  reminderDaysOfWeek: "1111111",
  updatedAt: "",
};

export function getNotificationSettings(): Promise<NotificationSettings> {
  return invoke<NotificationSettings>("get_notification_settings");
}

export function updateNotificationSettings(
  input: UpdateNotificationSettingsInput,
): Promise<NotificationSettings> {
  return invoke<NotificationSettings>("update_notification_settings", {
    input,
  });
}

export function evaluateReminders(
  input?: EvaluateRemindersInput,
): Promise<ReminderEvaluation> {
  return invoke<ReminderEvaluation>("evaluate_reminders", {
    input: input ?? null,
  });
}

export function dismissInAppReminder(reminderId: number): Promise<void> {
  return invoke<void>("dismiss_in_app_reminder", { reminderId });
}

export function sendTestNotification(): Promise<NotificationDispatchResult> {
  return invoke<NotificationDispatchResult>("send_test_notification");
}

File: apps/desktop/src-tauri/src/dto/notifications.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct NotificationSettingsDto {
    pub user_id: i64,
    pub notification_enabled: bool,
    pub in_app_reminders_enabled: bool,
    pub due_review_notifications_enabled: bool,
    pub streak_notifications_enabled: bool,
    pub reminder_time: String,
    pub reminder_days_of_week: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateNotificationSettingsDto {
    pub notification_enabled: bool,
    pub in_app_reminders_enabled: bool,
    pub due_review_notifications_enabled: bool,
    pub streak_notifications_enabled: bool,
    pub reminder_time: String,
    pub reminder_days_of_week: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct EvaluateRemindersInputDto {
    pub local_date: Option<String>,
    pub local_time: Option<String>,
    pub force: Option<bool>,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct InAppReminderDto {
    pub id: i64,
    pub kind: String,
    pub title: String,
    pub body: String,
    pub action_label: Option<String>,
    pub route: Option<String>,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ReminderEvaluationDto {
    pub reminders: Vec<InAppReminderDto>,
    pub new_reminders: Vec<InAppReminderDto>,
    pub due_review_count: i64,
    pub daily_goal_cards: i64,
    pub today_cards_reviewed: i64,
    pub current_streak: i64,
    pub settings: NotificationSettingsDto,
    pub evaluated_at: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct NotificationDispatchResultDto {
    pub reminder: InAppReminderDto,
    pub os_status: String,
    pub message: String,
}
